### ISSUE TITLE
Incorrect README displayed when GitHub repository linked as homepage

### ISSUE THREAD

**Issue by bhariharan** [npm/registry-issue-archive#50]
My package has 2 readme files - one for npm called `npm.md`, and another for my GitHub repository, called `README.md`. When I package my app and publish it, I have a script that renames these readme files, so `npm.md` becomes `README.md` and the original `README.md` becomes `ORIGINAL.md`. npm should display the `README.md` in my package, as opposed to displaying the `README.md` from my GitHub repo, since they're different in this case. This was working correctly until about 6 months ago. My packages where I see this issue are:
- [forcedroid](https://www.npmjs.com/package/forcedroid)
- [forceios](https://www.npmjs.com/package/forceios)

I had originally filed this issue [here](https://github.com/npm/www/issues/20), but was asked to re-file it here.

**Comment by bhariharan** [comment]
One approach to solve this problem would be to allow me to explicitly specify a `README` to use in `package.json`.

**Comment by bhariharan** [comment]
Any update here?

**Comment by soldair** [comment]
hey i forwarded the question of how best to direct you to the cli team.

im not sure why your prepublish script is not working the way you expect. 

sorry for the delay!

**Comment by bhariharan** [comment]
@soldair The `prepublish` script is working fine - I have validated that. Also, the package has the correct `README` - I have validated that as well. The problem is `npmjs.org` is not displaying the `README` from the package, it's displaying the `README` from `GitHub`, which is different.

**Comment by soldair** [comment]
hmm im sure there is a bug somewhere but we don't make external calls to services other than the registry to get the readme.

Now the registry does have 2 copies of your readme. One inside the tarball and the other inside the package metadata sent to us on publish via HTTP POST from the cli. What i think is that the version in the package metadata (the one that populates the website) has the unexpected value. This is probably due to a race condition in triggering prepublish scripts, read-package-json, and bundling the files.

could you confirm if this readme in the metadata is the version from github or the version for npm?

```
 curl https://registry.npmjs.com/forcedroid | json readme
```

**Comment by bhariharan** [comment]
@soldair Yes, you're right! The `readme` in the metadata contains the one from `GitHub`, not from the package. What's the right fix for this?

**Comment by bhariharan** [comment]
Is there a way to explicitly set `readme` in `package.json`? That would fix this, wouldn't it?

**Comment by soldair** [comment]
i looked at read-package-json which npm uses to build the metadata its going to send and if the readme value is populated it looks like thats the value of the readme. 

https://github.com/npm/read-package-json/blob/master/read-json.js#L205

I dont think you can modify the package.json in a hook. but if when ever the readme was changed you updated the package.json readme property with it's contents that would work. =/

 You could always make an publish-to-npm.sh that moves the files and runs npm publish instead of mucking about in race conditions due to running the sh as a prepublish npm script. It does sound like a bug that you cannot modify your readme in a prepublish script. Maybe if we go through cli issues there is a good reason.

**Comment by bhariharan** [comment]
@soldair Thanks for digging into this. I'll try to use a script that does this sequentially to avoid the race condition. Is there a way to validate this locally using `npm pack` instead of `npm publish`? Can I view the metadata generated and the package contents locally?

Also, would you recommend that I file this bug in the CLI repo?

**Comment by soldair** [comment]
yeah making a cli issue may be helpful at fixing or clarifying the actual race condition. 

if you run npm pack you can verify that the correct readme.md file is the one you are publishing in the module tarball itself, but this doesnt necessarily mean it's the one in the metadata. 

I cant really point you in a safe direction aside form making a wrapper .sh file to mutate the module contents then run `npm publish` and mutate the contents back. This script would be easy to test because you can comment out publish and verify that the README.md is correct.
